### PR TITLE
Issue 3167: Missing values that are NaNs are not supported

### LINKED ISSUE #3167: Missing values that are NaNs are not supported
It is possible to use NaNs as a missing value. Hence, Grid::GetMissingValue() may return a NaN. Much of the code in VAPOR that handles missing values checks for the existence of missing values in the data with something like:

`if (g->GetMissingValue() == g->GetValue())`

However, in C/C++ comparing a NaN to a NaN with an equality operator always results in False. Thus special case handling should be added throughout the code base to deal with NaN missing values. E.g.

`if ((g->GetMissingValue() == g->GetValue())) || (std::isnan(g->GetMissingValue())  && std::isnan(g->GetValue()))`

See the test data set used in #3087

### PR DESCRIPTION
**Pull request by clyne**
Fixed #3167 

This PR fixes #3167. However, the fix in this PR is internally replace NaN's with std::infinity. This is a much simpler fix than proposed in #3167, only touching two files, but may not prove to be as robust.

Note, this PR also fixes an unreported bug: The NetCDFCFCollection class did not handle the case where a single NetCDF
file used by the "_FillValue" and "missing_value" attribute name to indicate missing values.

### CODE REVIEW

**Comment by sgpearse** [comment]
@clyne - Can we make this a normal PR instead of a draft at this point?

**Comment by clyne** [comment]
> @clyne - Can we make this a normal PR instead of a draft at this point?

Indeed. Thanks for catching. Done.

File: lib/vdc/GridHelper.cpp [hunk 1]
```diff
@@ -606,6 +606,11 @@ void GridHelper::_makeGridHelper(const DC::DataVar &var, const DimsType &roi_dim
         mv = var.GetMissingValue();
     }
 
+    // Don't allow NaNs as missing values. We replace any
+    // missing values in the data with infinity
+    //
+    if (std::isnan(mv)) mv = std::numeric_limits<float>::infinity();
```
Comment (shaomeng) [inline lib/vdc/GridHelper.cpp:612]: I assume there won't be incoming data sets with double type and also containing missing values? 

Also, NetCDFCFCollection::_GetMissingValue() seems to represent mv as a double. So maybe there needs some consistency improvement, or better, complete coverage improvement (i.e., properly handle both float and double types)?

File: lib/vdc/NetCDFCFCollection.cpp [hunk 2]
```diff
@@ -826,55 +824,24 @@ bool NetCDFCFCollection::_GetMissingValue(string varname, string attname, double
...
+    // for either, but give priority to "missing_value"
     //
     for (int d = 1; d < 5; d++) {
         vector<string> vars = NetCDFCFCollection::GetDataVariableNames(d, false);
 
         for (int i = 0; i < vars.size(); i++) {
             double mv;
-            if (_GetMissingValue(vars[i], mv_attname, mv)) { missingValueMap[vars[i]] = mv; }
+            if (_GetMissingValue(vars[i], "missing_value", mv)) { missingValueMap[vars[i]] = mv; }
+            if (_GetMissingValue(vars[i], "_FillValue", mv)) { missingValueMap[vars[i]] = mv; }
```
Comment (shaomeng) [inline lib/vdc/NetCDFCFCollection.cpp:844]: if "missing_value" has a higher priority as the comment says, shouldn't these two lines reverse order?

**Review by sgpearse — CHANGES_REQUESTED** [review]
The fix works and I thought the code looked good, but Sam's raising a couple points that we should agree on.

File: lib/vdc/GridHelper.cpp (hunk 1 above)
Comment (clyne) [reply lib/vdc/GridHelper.cpp:612]: > I assume there won't be incoming data sets with double type and also containing missing values?

That is correct. The Grid class is not templated; data values are always single precision floats.

> 
> Also, NetCDFCFCollection::_GetMissingValue() seems to represent mv as a double. So maybe there needs some consistency improvement, or better, complete coverage improvement (i.e., properly handle both float and double types)?

Agreed. But I'm not sure that it is in scope for this fix.

File: lib/vdc/NetCDFCFCollection.cpp (hunk 2 above)
Comment (clyne) [reply lib/vdc/NetCDFCFCollection.cpp]: You are correct. Fixed. Thanks!

**Review by sgpearse — APPROVED** [review]

**Review by shaomeng — APPROVED** [review]

**Pull request merged**